lib/linguist/heuristics.rb

```diff
@@ -380,6 +380,26 @@ def call(data)
       end
     end
 
+    disambiguate ".pod" do |data|
+      if /^\s*=\w+$/.match(data)
+        if /^=pod|=cut/.match(data)
+          Language["Pod"]
+        elsif /^\s+=(begin|cut|pod)/.match(data)
```

Comment:
Hum, [not really](https://github.com/perl6/doc/blob/master/doc/Language/5to6-nutshell.pod6).

Reply:
@lildude A Pod command with leading whitespace is invalid in Perl/Pod 5, but valid in Perl/Pod 6. Therefore, if spaces *are* matched, it's unambiguously Pod 6.

Reply:
Oh. I didn't get that from the doc. TIL. In that case, I best change my comment :-)

Reply:
If I understand what's above (and my knowledge of Ruby is limited), it says that if it starts with `=pod`, it's necessarily Pod, but that's not the case. It could still be Pod 6. Maybe if you invert the condition, that would be true. If it's got a whitespace before `=pod`, it's pod 6. If it does not, it could be either.

Reply:
Yep, the first expression *will* match both Pod 5 and 6, so I'd move the second expression (which can only match Pod 6) to the top so it's evaluated first.

@JJ What you're reading is a regular expression with syntax [almost identical](https://ruby-doc.org/core-2.5.0/Regexp.html) to Perl's regex syntax (engine-specific extensions notwithstanding). =) The main difference in this code is that Ruby implicitly returns the value of the last evaluated expression in a function body (making the lines inside the conditionals equivalent to writing this:

~~~ruby
return Language["Pod 6"]
~~~

Reply:
**Python folk:** *"There should only be one obvious way to do it."*

**Perl folk:** *"There's more than one way to do it."*

**Ruby folk:** *"There's a shorter way to do it."*

Reply:
What the hell. How long has this mistake been there for...?

~~~ruby
/^=pod|=cut/
~~~

That should be:

~~~ruby
/^=(pod|cut)/
~~~

Because `/^=pod|=cut/` is currently matching:
* `^=pod` (`=pod` at the start of the line)
OR
* Any line containing `=cut`

That's... not your fault, but it needs to be fixed. :|

Reply:
> The main difference in this code is that Ruby implicitly returns the value of the last evaluated expression in a function body

For the record, so does Perl (and I assume Perl 6) ;)

The `=pod`/`=cut` heuristic could actually be dropped entirely, since it will already classify any document without indented directives or `=begin pod` or `=comment` as Pod.

Reply:
> For the record, so does Perl (and I assume Perl 6) ;)

Well, I'll be fucked, it does too. :|

HAHA, SEE?! Ruby pretty much stole everything from Perl, right down to the `__END__` directive and all. Pity that they could've learned a few pointers about dependency management in the RubyGems arena. 😆

`#TeamPerl5`

Reply:
> use \h+/\h* which only matches horizonal spaces

Nope. In Oniguruma (Ruby's regular expression engine), `\h` means to match a hexadecimal digit (saving you the effort of writing out `[A-Fa-f0-9]+` instead.

Admittedly, both classes have their uses, and their conflicts are a frequent cause of confusion by grammar authors... :(

Reply:
> Oh... are these regexes being applied to the whole file rather than a single line?

It calls `.match(data)` on the regex, and `data` contains the entire file from what I can tell.

Reply:
However, you can use the `\A` escape instead to anchor a match at the start of the entire string, multiline modes and `^` anchors be damned. =)

Reply:
Other options would be `[[:blank:]]` if it supports POSIX character classes, or just `[ \t]` (note the literal space) since tab and space are the only actual horizontal space characters.

Reply:
Unfortunately the anchors we want here are those of each line, not the beginning/end of the entire string, so we want multiline mode if that's possible.

Reply:
Ah, right. Don't forget you can still match horizontal whitespace characters directly using the `\p{XPosixBlank}` Unicode property escape. =)

Reply:
> Don't forget you can still match horizontal whitespace characters directly using the \p{XPosixBlank} Unicode property escape.

Ruby seems to disagree here.

```
/home/tyil/projects/personal/linguist/lib/linguist.rb:4:in `require': /home/tyil/projects/personal/linguist/lib/linguist/heuristics.rb:384: invalid character property name {XPosixBlank}: /^\p{XPosixBlank}*=\w+$/ (SyntaxError)
invalid character property name {XPosixBlank}: /^\p{XPosixBlank}+=(begin|cut|pod)/
invalid character property name {XPosixBlank}: /^\p{XPosixBlank}*=(comment|begin pod)/
        from /home/tyil/projects/personal/linguist/lib/linguist.rb:4:in `<top (required)>'
        from /home/tyil/projects/personal/linguist/test/helper.rb:4:in `require'
        from /home/tyil/projects/personal/linguist/test/helper.rb:4:in `<top (required)>'
        from /home/tyil/projects/personal/linguist/test/test_blob.rb:1:in `require_relative'
        from /home/tyil/projects/personal/linguist/test/test_blob.rb:1:in `<top (required)>'
        from /home/tyil/.gem/ruby/2.3.0/gems/rake-12.3.1/lib/rake/rake_test_loader.rb:17:in `require'
        from /home/tyil/.gem/ruby/2.3.0/gems/rake-12.3.1/lib/rake/rake_test_loader.rb:17:in `block in <main>'
        from /home/tyil/.gem/ruby/2.3.0/gems/rake-12.3.1/lib/rake/rake_test_loader.rb:5:in `select'
        from /home/tyil/.gem/ruby/2.3.0/gems/rake-12.3.1/lib/rake/rake_test_loader.rb:5:in `<main>'
rake aborted!
```

> or just [ \t]

This works. Can someone confirm that these are truly the only horizontal space characters in Unicode? @samcv ? Considering Perl 6 heavily promotes Unicode, I want to be sure it gets handled correctly for POD as well.

Reply:
It isn't, for example http://www.fileformat.info/info/unicode/char/2002/index.htm, which is matched by both `\h` and `[[:blank:]]` in Perl. I was mislead by the POSIX character class description regarding that...

Reply:
JFC... of course, I forgot about *another* common cause for Perl VS Oniguruma conflict. 😠Different syntax for Unicode property names, as you do. @Tyil, check [`doc/UNICODE_PROPERTIES`](https://github.com/kkos/oniguruma/blob/master/doc/UNICODE_PROPERTIES) in the official Oniguruma repository. It contains an index of supported unicode properties with their identifiers.

> Can someone confirm that these are truly the only horizontal space characters in Unicode?

Yes, that's why I'm helping you. Em-space, en-space, figure space, three-per-em space... check the entire ["Space Separator"](http://graphemica.com/categories/space-separator) block, for example.

But `[ \t]` alone is, as you correctly guessed, too naive if Unicode should be considered. ;)

Reply:
http://ruby-doc.org/core-2.5.0/Regexp.html#class-Regexp-label-Character+Classes seems to confirm two things: `[[:blank:]]` should be supported and include unicode characters, and `/m` should work to enable multiline mode.

Reply:
I take that second part back. Confusingly it looks like Ruby's multiline mode is equivalent to Perl's single-line mode (`/s` which enables `.` to match newlines) and that anchors match lines by default. So no modifier should be needed.

Reply:
This is what I would use: `[\s&&[^\n]]`. So any whitespace except a newline. This will make sure it allows the all the other control characters that can act as spaces. This works in my testing even when unusual things like \v \f and \r before the pod.

Reply:
I recall being bitten by that discrepancy numerous times. You're not alone... 😠

This is why I prefer Perl 5. 😉

Reply:
Sadly it does not look like any official unicode property will be helpful here if `[[:blank:]]` doesn't include unicode. XPosixBlank is a Perl extension.

Reply:
> Yes, that's why I'm helping you.

Much appreciated. From the document you linked, I found 4 candidates:

- 175: Pattern_White_Space
- 230: White_Space
- 237: Space_Separator
- 230: WSpace

However, @samcv's solution also looks workable for me. I'll try that one first, as I'm not familiar enough with Unicode to know if any of these candidates matches all the horizontal whitespace characters.

> [[:blank:]] should be supported and (possibly?) include unicode characters

I asked on `#ruby` on Freenode, and got the following response:

```
23:53 <apeiros> tyil: [[:blank:]] as per those docs is only space or tab, i.e. \x20 and \x09
```

So I guess we'll have to assume `[[:blank:]]` does not what we want to match.

Reply:
That's interesting @samcv that you can do character class intersections, but that one's not quite enough, there are other vertical whitespace characters too (of course!)

Reply:
@Tyil we don't need to use Unicode properties here. There isn't any property which does what we want because tabs and formfeed are not spaces they are control characters, i.e. they are not characters which separate words. Spaces have semantic meaning while tabs, newline, formfeed etc. do not; they and are formatting characters.